File: api/internal/app/model/date.go
package model

import (
	"time"
)

type (
	Year  int32
	Month int32
	Day   int32
)

type Date struct {
	Year  Year
	Month Month
	Day   Day
}

type YearMonth struct {
	Year  Year
	Month Month
}

func NewYearMonth(year Year, month Month) YearMonth {
	return YearMonth{
		Year:  year,
		Month: month,
	}
}

func NewYearMonthFromTime(t time.Time) YearMonth {
	return NewYearMonth(
		Year(t.Year()),
		Month(t.Month()),
	)
}

File: frontend/lib/entities/date.dart
import 'package:freezed_annotation/freezed_annotation.dart';
part 'date.freezed.dart';
part 'date.g.dart';

@freezed
class Date with _$Date {
  Date._();
  const factory Date({
    required int year,
    required int month,
    required int day,
  }) = _Date;

  factory Date.fromJson(Map<String, dynamic> json) => _$DateFromJson(json);
  factory Date.empty() => const Date(year: 0, month: 0, day: 0);

  bool get isZero => year == 0 && month == 0 && day == 0;
}

@freezed
class YearMonth with _$YearMonth {
  YearMonth._();
  const factory YearMonth({
    required int year,
    required int month,
  }) = _YearMonth;

  factory YearMonth.fromJson(Map<String, dynamic> json) =>
      _$YearMonthFromJson(json);

  factory YearMonth.empty() => const YearMonth(year: 0, month: 0);

  bool get isZero => year == 0 && month == 0;
}
